File: Java/Hasta Entraga 3/irrgarten/src/irrgarten/Labyrinth.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

import java.util.ArrayList;

/**
 *
 * @author alejandro
 */
public class Labyrinth {
    // Atributos de clase estáticos
    private static final char BLOCK_CHAR = 'X';
    private static final char EMPTY_CHAR = '-';
    private static final char MONSTER_CHAR = 'M';
    private static final char COMBAT_CHAR = 'C';
    private static final char EXIT_CHAR = 'E';
    private static final int ROW = 0;
    private static final int COL = 1;
    
    // Atributos de instancia
    private int nRows;
    private int nCols;
    private int exitRow;
    private int exitCol;
    
    // Tablas
    private Monster monster[][];
    private Player players[][];
    private char laberinto[][];
    
    // Constructor
    public Labyrinth(int nRows, int nCols, int exitRow, int exitCol){
        this.nRows = nRows;
        this.nCols = nCols;
        this.exitRow = exitRow;
        this.exitCol = exitCol;
        
        this.monster = new Monster[nRows][nCols];
        this.players = new Player[nRows][nCols];
        this.laberinto = new char[nRows][nCols];
        
        //Inicializo el tablero de caracteres
        for(int i=0; i <nRows;i++){
            for(int j=0; j < nCols;j++){
                this.laberinto[i][j] = EMPTY_CHAR;
            }
        }
        
        //Coloco la salida del laberinto
        this.laberinto[exitRow][exitCol] = EXIT_CHAR;
    }
    
    // Método completado
    public void spreadPlayers(ArrayList<Player> players){
        for(int i = 0; i < players.size();i++){
            Player p = players.get(i);
            
            int pos[] = randomEmptyPos();
            
            putPlayer2D(-1,-1,pos[ROW],pos[COL],p);
        }
    }
    
    // Devuelve true si el jugador está en la casilla de salida
    public boolean haveAWinner(){
        return this.players[this.exitRow][this.exitCol] != null;
    }
    
    //
    @Override
    public String toString(){
        String s = " ";
        
        for(int i=0; i < nRows;i++){
            for(int j=0; j < nCols;j++){
                s = s + this.laberinto[i][j] + "\t";
            }
            s = s + "\n";
        }
        return s;
    }
    
    // Comprueba si la posición dada está dentro del tablero y si lo es
    // añade al monstruo al laberinto y actualiza su posición
    public void addMonster(int row, int col, Monster monstruo){
        if(posOK(row,col) && emptyPos(row,col)){
            this.laberinto[row][col] = MONSTER_CHAR;
            
            this.monster[row][col] = monstruo;
            monstruo.setPos(row,col);
        }
    }
    
    // Método completado
    public Monster putPlayer(Directions direction, Player player){
        int oldRow = player.getRow();
        int oldCol = player.getCol();
        
        int newPos[] = dir2Pos(oldRow,oldCol,direction);
        Monster monster = putPlayer2D(oldRow,oldCol,newPos[ROW], newPos[COL],player);
        
        return monster;
    }
    
    
    // Método completado
    public void addBlock(Orientation orientation, int startRow, int startCol, int length){
        int incRow;
        int incCol;
        
        if(orientation == Orientation.VERTICAL){
            incRow = 1;
            incCol = 0;
        } else{
            incRow = 0;
            incCol = 1;
        }
        
        int row = startRow;
        int col = startCol;
        
        while((posOK(row,col)) && (length > 0)){
            laberinto[row][col] = BLOCK_CHAR;
            length -= 1;
            row += incRow;
            col += incCol;
        }
    }
    
    // Método completado
    public ArrayList<Directions> validMoves(int row,int col){
        ArrayList<Directions> output = new ArrayList<>();
        
        if(canStepOn(row+1,col)){
            output.add(Directions.DOWN);
        }
        if(canStepOn(row-1,col)){
            output.add(Directions.UP);
        }
        if(canStepOn(row,col+1)){
            output.add(Directions.RIGHT);
        }
        if(canStepOn(row, col-1)){
            output.add(Directions.LEFT);
        }
        return output;
    }
    
    //
    private boolean posOK(int row,int col){
        return (row >= 0 && row < nRows) && (col >= 0 && col < nCols);
    }
    
    // 
    private boolean emptyPos(int row, int col){
        return this.laberinto[row][col] == EMPTY_CHAR;
    }
    
    // Devuelve true si la posición suministrada alberga solo un monstruo
    private boolean monsterPos(int row, int col){
        return this.laberinto[row][col] == MONSTER_CHAR;
    }
    
    // Devuelve true si la posición suministrada es la de salida
    private boolean exitPos(int row, int col){
        return row == this.exitRow && col == this.exitCol;
    }
    
    // Devuelve true si la posición suministrada contiene a la vez un monstruo y un jugador
    private boolean combatPos(int row,int col){
        return this.laberinto[row][col] == COMBAT_CHAR;
    }
    
    // Indica si la posición suministrada está dentro del laberinto y se
    // corresponde con una de estas tres opciones: casilla vacía, casilla donde 
    // habita un monstruo o salida
    private boolean canStepOn(int row, int col){
        return posOK(row,col) && (emptyPos(row,col) || monsterPos(row,col) || exitPos(row,col));
    }
    
    // Actualiza la casilla que un jugador abandona
    private void updateOldPos(int row, int col){
        if(posOK(row,col)){
            if(combatPos(row,col)){
                this.laberinto[row][col] = MONSTER_CHAR;
            } else{
                this.laberinto[row][col] = EMPTY_CHAR;
            }
        }
    }
    
    // Calcula la nueva posición [row, col] basada en una dirección 
    private int[] dir2Pos(int row, int col,Directions direction){
        int[] newPos = {row,col};
        
        // Switch con reglas, cambiado porque netbeans me daba la opción
        switch (direction){
            case UP -> newPos[ROW]--;
            case DOWN -> newPos[ROW]++;
            case LEFT -> newPos[COL]--;
            case RIGHT -> newPos[COL]++;
        }
        
        return newPos;
    }
    
    // Genera una posición [row, col] aleatoria que esté vacía 
    private int[] randomEmptyPos(){
        int row, col;
        
        do{
            row = Dice.randomPos(this.nRows);
            col = Dice.randomPos(this.nCols);
        } while (!this.emptyPos(row, col));
        
        return new int[]{row,col};
    }
    
    // Método completado
    private Monster putPlayer2D(int oldRow, int oldCol, int row, int col, Player player){
        Monster output = null;
        
        if(canStepOn(row,col)){
            if(posOK(oldRow,oldCol)){
                Player p = players[oldRow][oldCol];
                if(p == player){
                    updateOldPos(oldRow,oldCol);
                    players[oldRow][oldCol] = null;
                }
            }
            
            boolean monsterPos = monsterPos(row,col);
            
            if(monsterPos){
                laberinto[row][col] = COMBAT_CHAR;
                output = monster[row][col];
            }else{
                char number = (char)('0' + player.getNumber());
                laberinto[row][col] = number;          
            }
            
            players[row][col] = player;
            player.setPos(row,col);
        }
        
        return output;
    }
}

File: Ruby/Hasta Entrega 3/Irrgarten/Irrgarten/labyrinth.rb
require_relative 'directions'

module Irrgarten
  class Labyrinth
    # Atributos de clase
    @@block_char = 'X'
    @@empty_char = '-'
    @@monster_char = 'M'
    @@combat_char = 'C'
    @@exit_char = 'E'
    @@row = 0
    @@col = 1

    def initialize(n_row,n_col,exit_row,exit_col)
      # Atributos de instancia
      @n_row = n_row
      @n_col = n_col
      @exit_row = exit_row
      @exit_col = exit_col

      # Tablas
      @laberinto = Array.new(@n_row) { Array.new(@n_col,) }
      @monster = Array.new(@n_row) { Array.new(@n_col) }
      @players = Array.new(@n_row) { Array.new(@n_col) }

      # Inicializo el tablero

      (0...@n_row).each do |i|
        (0...@n_col).each do |j|
          @laberinto[i][j] = @@empty_char # Usando la constante
        end
      end

      # Coloco la salida del laberinto
      @laberinto[@exit_row][@exit_col] = @@exit_char
    end

    # Metodo completado
    def spread_players(players)
      players.each do |p|
        pos = random_empty_pos
        put_player_2d(-1,-1,pos[@@row],pos[@@col],p)
      end
    end

    # Devuelve true si el jugador está en la casilla de salida
    def have_a_winner
      @players[@exit_row][@exit_col] != nil
    end

    # Método to_string
    def to_s
      s = " "

      (0...@n_row).each do |i|
        (0...@n_col).each do |j|
          s += @laberinto[i][j]
        end
        s += "\n"
      end
      s
    end

    # Comprueba si la posición dada está dentro del tablero y si lo es
    # añade al monstruo al laberinto y actualiza su posición
    def add_monster(row, col, monstruo)
      if pos_ok(row, col) && empty_pos(row, col)
        @laberinto[row][col] = @@monster_char
        @monster[row][col] = monstruo
        monstruo.set_pos(row, col)
      end
    end

    # Metodo completado
    def put_player(direction, player)
      old_row = player.row
      old_col = player.col

      new_pos = dir2_pos(old_row, old_col, direction)

      monster = put_player_2d(old_row, old_col, new_pos[@@row], new_pos[@@col], player)

      monster
    end

    # Metodo completado
    def add_block(orientation, start_row, start_col, length)
      if orientation == Orientation::VERTICAL
        inc_row = 1
        inc_col = 0
      else
        inc_row = 0
        inc_col = 1
      end

      row = start_row
      col = start_col

      while pos_ok(row, col) && (length > 0)
        @laberinto[row][col] = @@block_char
        length -= 1
        row += inc_row
        col += inc_col
      end
    end

    # Metodo completado
    def valid_moves(row, col)
      output = [] # Array vacío

      # Añadimos direcciones si can_step_on es true
      # (Asumo que existe un módulo Directions::UP, ::DOWN, etc.)
      output << Directions::DOWN if can_step_on(row + 1, col)
      output << Directions::UP if can_step_on(row - 1, col)
      output << Directions::RIGHT if can_step_on(row, col + 1)
      output << Directions::LEFT if can_step_on(row, col - 1)

      output
    end

    # --- Métodos Privados ---
    private

    # Devuelve true si la posición proporcionada está dentro del laberinto
    def pos_ok(row, col)
      (row >= 0 && row < @n_row) && (col >= 0 && col < @n_col)
    end

    # Devuelve true si la posición suministrada está vacía
    def empty_pos(row, col)
      @laberinto[row][col] == @@empty_char
    end

    # Devuelve true si la posición suministrada alberga solo un monstruo
    def monster_pos(row, col)
      @laberinto[row][col] == @@monster_char
    end

    # Devuelve true si la posición suministrada es la de salida
    def exit_pos(row, col)
      row == @exit_row && col == @exit_col
    end

    # Devuelve true si la posición suministrada contiene un monstruo y un jugador
    def combat_pos(row, col)
      @laberinto[row][col] == @@combat_char
    end

    # Indica si la posición es válida y se puede pisar (vacía, monstruo o salida)
    def can_step_on(row, col)
      pos_ok(row, col) && (empty_pos(row, col) || monster_pos(row, col) || exit_pos(row, col))
    end

    # Actualiza la casilla que un jugador abandona
    def update_old_pos(row, col)
      if pos_ok(row, col)
        if combat_pos(row, col)
          @laberinto[row][col] = @@monster_char
        else
          @laberinto[row][col] = @@empty_char
        end
      end
    end

    # Calcula la nueva posición [row, col] basada en una dirección
    def dir2_pos(row, col, direction)
      new_pos = [row, col]

      case direction
      when Directions::UP
        new_pos[@@row] -= 1
      when Directions::DOWN
        new_pos[@@row] += 1
      when Directions::LEFT
        new_pos[@@col] -= 1
      when Directions::RIGHT
        new_pos[@@col] += 1
      end

      new_pos
    end

    # Genera una posición [row, col] aleatoria que esté vacía
    def random_empty_pos
      row = 0
      col = 0

      begin
        row = Dice.random_pos(@n_row)
        col = Dice.random_pos(@n_col)
      end until empty_pos(row, col)

      [row, col]
    end

    # Metodo completado
    def put_player_2d(old_row, old_col, row, col, player)
      output = nil  # Equivalente a 'output = null'

      if can_step_on(row, col)

        if pos_ok(old_row, old_col)
          p = @players[old_row][old_col]
          if p == player
            update_old_pos(old_row, old_col)
            @players[old_row][old_col] = nil
          end
        end

        is_monster_pos = monster_pos(row, col)

        if is_monster_pos
          @laberinto[row][col] = @@combat_char
          output = @monster[row][col]
        else
          number_char = player.number.to_s
          @laberinto[row][col] = number_char
        end

        @players[row][col] = player
        player.set_pos(row, col)
      end

      output
    end


  end
end